main.cpp

```diff
@@ -131,217 +115,145 @@ struct IntType;
 
-    FloatType& divide( float input )
+FloatType& FloatType::divide( float input )
+{
+    if (input == 0.f)
     {
-        if (input == 0.f)
-        {
-            std::cout << "warning: floating point division by zero!" << std::endl;
-        }
-        *value /= input;
-
-        return *this;
+        std::cout << "warning: floating point division by zero!" << std::endl;
     }
+    if (value != nullptr) *value /= input;
 
-    FloatType& divide( const FloatType& input ); 
-    FloatType& divide( const DoubleType& input );
-    FloatType& divide( const IntType& input );
-};
+    return *this;
+}
 
 struct DoubleType
 {
-    double* value = nullptr;
     DoubleType(double val) : value(new double(val)) {}
     ~DoubleType( )
     {
         delete value;
     }
+    operator double() { return *value; }
 
-    DoubleType& add ( double input )
-    {
-        *value += input; 
-        return *this;
-    }
-
-    DoubleType& subtract ( double input )
-    {
-        *value -= input;
-        return *this;
-    }
+    DoubleType& add ( double input );
+    DoubleType& subtract ( double input );
+    DoubleType& multiply ( double input );
+    DoubleType& divide ( double input );
 
-    DoubleType& multiply ( double input )
-    {
-        *value *= input; 
-        return *this;
-    }
-
-    DoubleType& divide ( double input )
-    {
-        if (input == 0.)
-        {
-            std::cout << "warning: floating point division by zero!" << std::endl;
-        }
-        *value /= input; 
-        return *this;
-    }
-
-    DoubleType& add( const FloatType& input ); 
-    DoubleType& add( const DoubleType& input );
-    DoubleType& add( const IntType& input );
-
-    DoubleType& subtract( const FloatType& input ); 
-    DoubleType& subtract( const DoubleType& input );
-    DoubleType& subtract( const IntType& input );
-
-    DoubleType& multiply( const FloatType& input ); 
-    DoubleType& multiply( const DoubleType& input );
-    DoubleType& multiply( const IntType& input );
-    
-    DoubleType& divide( const FloatType& input ); 
-    DoubleType& divide( const DoubleType& input );
-    DoubleType& divide( const IntType& input );
+private:
+    double* value = nullptr;
 };
 
-struct IntType
+DoubleType& DoubleType::add ( double input )
 {
-    int* value;
-
-    IntType(int val) : value(new int(val)) {}
-    ~IntType() { delete value; }
+    if (value != nullptr) *value += input; 
+    return *this;
+}
 
-    IntType& add (int input)
-    {
-        *value += input;
-        return *this;
-    }
+DoubleType& DoubleType::subtract ( double input )
+{
+    if (value != nullptr) *value -= input;
+    return *this;
+}
 
-    IntType& subtract (int input)
-    {
-        *value -= input;
-        return *this;
-    }
+DoubleType& DoubleType::multiply ( double input )
+{
+    if (value != nullptr) *value *= input; 
+    return *this;
+}
 
-    IntType& multiply (int input)
+DoubleType& DoubleType::divide ( double input )
+{
+    if (input == 0.)
```

Comment:
while this isn't wrong, it's generally not as readable as writing
`if (input == 0.0)`

it's easy to miss that `.`
- [ ] revise